### ISSUE TITLE
[v0.51.0] hab pkg binlink is not symlinking my bin

### ISSUE THREAD

**Issue by bdangit** [habitat-sh/habitat#4390]
It appears `hab pkg binlink`ing is not working as expected as it appears to symlink anything it can find in `$PATH`.

Here is my binary as built.
```
[79][default:/src:0]# ls $(hab pkg path bdangit/cpp-gtest-demo)/bin
Tutorial  libMathFunctions.a
[80][default:/src:0]# ls $(hab pkg path bdangit/cpp-gtest-demo)
[80][default:/src:0]# cd $(hab pkg path bdangit/cpp-gtest-demo)
[81][default:/hab/pkgs/bdangit/cpp-gtest-demo/0.1.0/20171226230643:0]# cd bin
[82][default:/hab/pkgs/bdangit/cpp-gtest-demo/0.1.0/20171226230643/bin:0]# ls
Tutorial  libMathFunctions.a
[83][default:/hab/pkgs/bdangit/cpp-gtest-demo/0.1.0/20171226230643/bin:0]# ldd Tutorial
        linux-vdso.so.1 (0x00007ffc099ee000)
        libstdc++.so.6 => /hab/pkgs/core/gcc-libs/5.2.0/20170513212920/lib/libstdc++.so.6 (0x00007fe42dbca000)
        libm.so.6 => /hab/pkgs/core/glibc/2.22/20170513201042/lib/libm.so.6 (0x00007fe42d8cc000)
        libgcc_s.so.1 => /hab/pkgs/core/gcc-libs/5.2.0/20170513212920/lib/libgcc_s.so.1 (0x00007fe42d6b6000)
        libc.so.6 => /hab/pkgs/core/glibc/2.22/20170513201042/lib/libc.so.6 (0x00007fe42d312000)
        /hab/pkgs/core/glibc/2.22/20170513201042/lib/ld-linux-x86-64.so.2 (0x00007fe42df44000)
[84][default:/hab/pkgs/bdangit/cpp-gtest-demo/0.1.0/20171226230643/bin:0]# ./Tutorial
./Tutorial Version 1.0
Usage: ./Tutorial number
```

When I try to binlink, I should only expect to have symlinks for `Tutorial`; instead I get the following:

```
[73][default:/src:1]# hab pkg binlink bdangit/cpp-gtest-demo
» Binlinking pldd from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked pldd from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/pldd
» Binlinking gencat from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked gencat from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/gencat
» Binlinking zic from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked zic from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/zic
» Binlinking iconvconfig from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked iconvconfig from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/iconvconfig
» Binlinking catchsegv from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked catchsegv from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/catchsegv
» Binlinking nscd from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked nscd from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/nscd
» Binlinking localedef from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked localedef from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/localedef
» Binlinking iconv from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked iconv from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/iconv
» Binlinking getent from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked getent from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/getent
» Binlinking rpcgen from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked rpcgen from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/rpcgen
» Binlinking ldd from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked ldd from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/ldd
» Binlinking locale from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked locale from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/locale
» Binlinking ldconfig from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked ldconfig from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/ldconfig
» Binlinking mtrace from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked mtrace from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/mtrace
» Binlinking tzselect from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked tzselect from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/tzselect
» Binlinking pcprofiledump from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked pcprofiledump from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/pcprofiledump
» Binlinking getconf from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked getconf from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/getconf
» Binlinking sprof from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked sprof from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/sprof
» Binlinking sotruss from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked sotruss from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/sotruss
» Binlinking zdump from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked zdump from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/zdump
» Binlinking makedb from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked makedb from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/makedb
» Binlinking xtrace from bdangit/cpp-gtest-demo into /hab/bin
★ Binlinked xtrace from bdangit/cpp-gtest-demo/0.1.0/20171226230643 to /hab/bin/xtrace
...
```

When I try to specify just `Tutorial`:

```
[76][default:/src:0]# hab pkg binlink -f bdangit/cpp-gtest-demo Tutorial
» Binlinking Tutorial from bdangit/cpp-gtest-demo into /hab/bin
✗✗✗
✗✗✗ `Tutorial' was not found under any 'PATH' directories in the bdangit/cpp-gtest-demo/0.1.0/20171226230643 package
✗✗✗
[77][default:/src:1]# echo $PATH
/hab/bin:/hab/pkgs/core/ncurses/6.0/20170513213009/bin:/hab/pkgs/core/glibc/2.22/20170513201042/bin:/hab/pkgs/core/vim/8.0.0004/20170513220742/bin:/hab/pkgs/core/attr/2.4.47/20170513213059/bin:/hab/pkgs/core/acl/2.2.52/20170513213108/bin:/hab/pkgs/core/util-linux/2.27.1/20170513214732/bin:/hab/pkgs/core/mg/20160118/20170513220846/bin:/hab/pkgs/core/make/4.2.1/20170513214620/bin:/hab/pkgs/core/less/481/20170513213936/bin:/hab/pkgs/core/pcre/8.38/20170513213423/bin:/hab/pkgs/core/diffutils/3.3/20170513214227/bin:/hab/pkgs/core/hab-plan-build/0.51.0/20171219022549/bin:/hab/pkgs/core/xz/5.2.2/20170513214327/bin:/hab/pkgs/core/wget/1.19.1/20171024102323/bin:/hab/pkgs/core/unzip/6.0/20170513215357/bin:/hab/pkgs/core/tar/1.29/20170513213607/bin:/hab/pkgs/core/sed/4.2.2/20170513213123/bin:/hab/pkgs/core/rq/0.9.2/20170612005822/bin:/hab/pkgs/core/hab/0.51.0/20171219021329/bin:/hab/pkgs/core/gzip/1.6/20170513214605/bin:/hab/pkgs/core/grep/2.22/20170513213444/bin:/hab/pkgs/core/gawk/4.1.3/20170513213646/bin:/hab/pkgs/core/findutils/4.4.2/20170513214305/bin:/hab/pkgs/core/file/5.24/20170513201915/bin:/hab/pkgs/core/coreutils/8.25/20170513213226/bin:/hab/pkgs/core/bzip2/1.0.6/20170513212938/bin:/hab/pkgs/core/binutils/2.25.1/20170513201927/bin:/hab/pkgs/core/bash/4.3.42/20170513213519/bin
```

**Comment by smacfarlane** [comment]
Possibly related:  https://github.com/habitat-sh/habitat/pull/3532

**Comment by smacfarlane** [comment]
@bdangit  I've dug into this a bit and it looks like there are two things going on here.  

The first is that there was a change to what a package considers its PATH, that affected how binlink works.  It used to use the PATH metafile, which only contained that packages `pkg_bin_dirs`. Now it uses RUNTIME_ENVIRONMENT which sets the PATH to include transient dependencies.   This had the side effect of causing binlink to bring in all the things. I think there are two things to be addressed here. The first is that binlink should be updated to report the correct package name that it's getting a binary from. The second would be to make that behavior opt-in/out,  so that you can get all the things in my package, or all the things my package might need to run. 

The second is it _looks_ like that when binlink tests to see if Tutorial is a file, it's failing which is very odd: https://github.com/habitat-sh/habitat/blob/master/components/core/src/fs.rs#L275-L300 .  I think that should bubble up a more accurate error, since there are ways that can fail that are not 'doesnt exist'.  You can see the same behavior with `core/aws-cli` and the __pycache__ directory.   

I think what I'd like to do here is to split this into three separate issues: source package in binlink ouput, opt-in/out for linking all the things, and better error reporting and handling when the target is not a file.  I can go ahead and do that, if you'd like.

**Comment by lilianmoraru** [comment]
Yep, this bug is very annoying :)
I wanted a few links(`hab pkg binlink be/gdb`) and it completely destroyed my `/usr/local/bin`(links overridden - 241 new links...).
Using binlink on an old build: `hab pkg binlink -d $(mktemp -d) core/file` correctly creates only the binaries in `/bin`(1 binary in this case).
On a newer build: `hab pkg binlink -d $(mktemp -d) be/file` - for this particular case I get 23 links...

**Comment by bdangit** [comment]
> I think what I'd like to do here is to split this into three separate issues: source package in binlink ouput, opt-in/out for linking all the things, and better error reporting and handling when the target is not a file. I can go ahead and do that, if you'd like.

@smacfarlane, yes, please! 

BTW, I added `pkg_bin_dirs=(bin)` and `Tutorial` was able to be binlinked. I didn’t think this was a necessary step but what you explained seems to be what’s going on.

**Comment by fnichol** [comment]
Sorry about being late here, but the good news is that I think this is now resolved with the work in master and the remainder will ship in the Habitat release (or currently in unstable builds). There's a bit more explanation in habitat-sh/builder#240 and the related work is in:

* habitat-sh/habitat#4722
* habitat-sh/habitat#4895
* habitat-sh/habitat#4888
* habitat-sh/core#19
* habitat-sh/core#20
* habitat-sh/core#21

As I'm pretty sure we've solved this now, I'm going to close this out. If I'm wrong--we can reopen.
